frontend/src/metabase/parameters/utils/dashboards.ts

```diff
@@ -0,0 +1,273 @@
+import { getParameterTargetField } from "metabase/parameters/utils/targets";
+import { isFieldFilterParameter } from "metabase/parameters/utils/parameter-type";
+import { slugify } from "metabase/lib/formatting";
+import {
+  UiParameter,
+  FieldFilterUiParameter,
+  ParameterWithTarget,
+} from "metabase/parameters/types";
+import {
+  ParameterId,
+  Parameter,
+  ParameterMappingOptions,
+  ParameterDimensionTarget,
+  ParameterVariableTarget,
+} from "metabase-types/types/Parameter";
+import {
+  Dashboard,
+  DashboardParameterMapping,
+  DashboardOrderedCard,
+} from "metabase-types/api";
+import { SavedCard, CardId } from "metabase-types/types/Card";
+import Metadata from "metabase-lib/lib/metadata/Metadata";
+import Field from "metabase-lib/lib/metadata/Field";
+
+type Mapping = DashboardParameterMapping & {
+  dashcard_id: number;
+  card: SavedCard;
+};
+
+export function createParameter(
+  option: ParameterMappingOptions,
+  parameters: Parameter[] = [],
+): Parameter {
+  let name = option.combinedName || option.name;
+  let nameIndex = 0;
+  // get a unique name
+  while (_.any(parameters, p => p.name === name)) {
+    name = (option.combinedName || option.name) + " " + ++nameIndex;
+  }
+
+  const parameter = {
+    name: "",
+    slug: "",
+    id: Math.floor(Math.random() * Math.pow(2, 32)).toString(16),
+    type: option.type,
+    sectionId: option.sectionId,
+  };
+  return setParameterName(parameter, name);
+}
+
+export function setParameterName(
+  parameter: Parameter,
+  name: string,
+): Parameter {
+  if (!name) {
+    name = "unnamed";
+  }
+  const slug = slugify(name);
+  return {
+    ...parameter,
+    name: name,
+    slug: slug,
+  };
+}
+
+export function setParameterDefaultValue(
+  parameter: Parameter,
+  value: any,
+): Parameter {
+  return {
+    ...parameter,
+    default: value,
+  };
+}
+
+export function hasMapping(parameter: Parameter, dashboard: Dashboard) {
+  return dashboard.ordered_cards.some(ordered_card => {
+    return ordered_card?.parameter_mappings?.some(parameter_mapping => {
+      return parameter_mapping.parameter_id === parameter.id;
+    });
+  });
+}
+
+export function isDashboardParameterWithoutMapping(
+  parameter: Parameter,
+  dashboard: Dashboard,
+) {
+  if (!dashboard) {
+    return false;
+  }
+
+  const parameterExistsOnDashboard = dashboard.parameters.some(
+    dashParam => dashParam.id === parameter.id,
+  );
+  const parameterHasMapping = hasMapping(parameter, dashboard);
+
+  return parameterExistsOnDashboard && !parameterHasMapping;
+}
+
+export function getDashboardUiParameters(
+  dashboard: Dashboard,
+  metadata: Metadata,
+): UiParameter[] {
+  const { parameters, ordered_cards } = dashboard;
+  const mappings = getMappings(ordered_cards);
+  const uiParameters: UiParameter[] = parameters.map(parameter => {
+    if (isFieldFilterParameter(parameter)) {
+      return buildFieldFilterUiParameter(parameter, mappings, metadata);
+    }
+
+    return {
```

Comment:
Every parameter is a `FieldFilterUiParameter` right now, but if we ever add new parameter types they will hit this `return`.